warp_drive/training/lightning-trainer.py

```diff
@@ -0,0 +1,520 @@
+                self.cuda_envs.cuda_data_manager,
+                action_name=_ACTIONS + suffix,
+                num_actions=action_dim[0],
+            )
+        else:
+            for action_idx, _ in enumerate(action_dim):
+                self.cuda_sample_controller.register_actions(
+                    self.cuda_envs.cuda_data_manager,
+                    action_name=f"{_ACTIONS}_{action_idx}" + suffix,
+                    num_actions=action_dim[action_idx],
+                )
+
+    def _evaluate_policies(self, batch_index=0):
+        """
+        Perform the policy evaluation (forward pass through the models)
+        and compute action probabilities
+        """
+        assert isinstance(batch_index, int)
+        probabilities = {}
+        for policy in self.policies:
+            probabilities[policy], _ = self.models[policy](
+                batch_index=batch_index, batch_size=self.training_batch_size_per_env
+            )
+
+        # Combine probabilities across policies if there are multiple policies,
+        # yet they share the same action placeholders.
+        # The action sampler will then need to run just once on each action type.
+        if (
+            len(self.policies) > 1
+            and not self.create_separate_placeholders_for_each_policy
+        ):
+            # Assert that all the probabilities are of the same length
+            # in other words the number of action types for each policy
+            # is the same.
+            num_action_types = {}
+            for policy in self.policies:
+                num_action_types[policy] = len(probabilities[policy])
+            assert all_equal(list(num_action_types.values()))
+
+            # Initialize combined_probabilities.
+            first_policy = list(probabilities.keys())[0]
+            num_action_types = num_action_types[first_policy]
+
+            first_action_idx = 0
+            num_envs = probabilities[first_policy][first_action_idx].shape[0]
+            num_agents = self.cuda_envs.env.num_agents
+
+            combined_probabilities = [None for _ in range(num_action_types)]
+            for action_type in range(num_action_types):
+                action_dim = probabilities[first_policy][action_type].shape[-1]
+                combined_probabilities[action_type] = torch.zeros(
+                    (num_envs, num_agents, action_dim)
+                ).cuda()
+
+            # Combine the probabilities across policies
+            for action_idx in range(num_action_types):
+                for policy, prob_values in probabilities.items():
+                    agent_to_id_mapping = self.policy_tag_to_agent_id_map[policy]
+                    combined_probabilities[action_idx][
+                        :, agent_to_id_mapping
+                    ] = prob_values[action_idx]
+
+            probabilities = {_COMBINED: combined_probabilities}
+
+        return probabilities
+
+    def forward(self, batch_index, start_event, end_event):
```

Comment:
The logic should be implemented in training_step, validation_step, test_step and predict_step.